rcl_lifecycle/include/rcl_lifecycle/rcl_lifecycle.h

```diff
@@ -247,6 +249,7 @@ rcl_lifecycle_state_machine_init(
   const rosidl_service_type_support_t * ts_srv_get_available_states,
   const rosidl_service_type_support_t * ts_srv_get_available_transitions,
   const rosidl_service_type_support_t * ts_srv_get_transition_graph,
+  bool enable_com_interface,
```

Comment:
So I guess my only real concern here is this change, which is an API break.  Looking at the development history of rcl/rcl_lifecycle over the last year, there are only 2 other API breaks that I see:

* #830, which switched an unsigned int -> uint8_t (so while technically an API break, unlikely to be noticed by most users)
* #676 , which moved the logging header file from `rcl` into `rcl_logging_interface` (so any users of the API would have to include it from somewhere else.  Its unlikely to be a problem anyway because it is really an internal API)

So this would be the first real API break we have.  We could avoid it by leaving this API as-is, marking it deprecated, and adding in `rcl_lifecycle_state_machine_init2`, or whatever.

@jacobperron @wjwwood @ivanpauno How do you feel about an API break here?

Reply:
What is wrong with an API change here? It is a new ROS distro and should be eligible for an API break in my mind. It's a bit unfortunate that we don't have function overloading in C, but renaming it to `..init2` seems pretty complicated to me as people then have to change it back to the original signature for H-Turtle.

Reply:
The churn of breaking API is that non-official client libraries have to catch up, e.g. rcljava.
But I haven't seem the lifecycle node interface wrapped in other client library that's not rclcpp.

This would also be a compile error, which is easy to detect and correct.

IMHO, this API break is reasonable.

---

We could also define a `rcl_lifecycle_state_machine_options_t`, so code will look like:

```c
rcl_lifecycle_state_machine_options_t options = rcl_lifecycle_get_default_state_machine_options();
rcl_ret_t ret = rcl_lifecycle_state_machine_init(
  state_machine,
  node_handle,
  <lots of typesupports>,
  options);
  ```

and next time a new option is added you only break ABI.
If a different client library doesn't care about exposing the new option they don't have to modify code, only re-compile.

Reply:
I like the idea of having a dedicated options struct. We could technically even turn this up a notch by introducing a `com_interface_options` struct which entails all the typesupports as well. That way, the function signature doesn't have millions of arguments ;-)

Reply:
> What is wrong with an API change here?

If this was a breaking change in rclcpp, we would most likely do the deprecation and adding of a new API.  We do try to do the tick-tock method of deprecation.

> The churn of breaking API is that non-official client libraries have to catch up, e.g. rcljava.

But given that this is in rcl (so isn't a user-facing API), we can more easily get away with an API break.  That being said, I know for sure that we will break rclnodejs: https://github.com/RobotWebTools/rclnodejs/blob/8490c2f8fa26019b8b284bcb17dc2001363d9195/src/rcl_lifecycle_bindings.cpp#L78

Still, if the maintainers of rcl think that is OK, it is OK with me.  If we go ahead with the API break, I will request a release note at https://docs.ros.org/en/rolling/Releases/Release-Galactic-Geochelone.html

Reply:
Breaking API has to happen sometimes, and in C it's less friendly, unfortunately, but I don't think we should ever avoid a change because it breaks API (assuming there's isn't a better way to do it). Without C++'s overloading, doing a deprecation cycle is annoying and not worth it much of the time. We have the versioning scheme that allows downstream stuff to `#ifdef` their way around it, e.g.:

https://github.com/ros2/rmw/blob/ace271650ad39a656d81239050c9639c4190f077/rmw/CMakeLists.txt#L81

I don't think we're doing that for `rcl` or `rcl_lifecycle` but we could.

Reply:
>Generic rcl_lifecycle_state_machine_options_t struct with a single parameter "enable_com_interface"

I was thinking in including the `default_states` and `allocator` arguments in the struct too, which have reasonable defaults.

```c
typedef struct rcl_lifecycle_state_machine_options_t {
  bool enable_com_interface;
  bool default_states;
  rcl_allocator_t allocator;
} rcl_lifecycle_state_machine_options_t;
```

> or more specific rcl_lifecycle_com_interface_options_t which has the parameter as well as the typesupports?

that also makes sense to me, the `rcl_lifecycle_state_machine_options_t` struct could also have the typesupports and `rcl_lifecycle_get_default_state_machine_options()` could provide the default ones (like [here](https://github.com/ros2/rcl/blob/5ef4b2dfca54c1a5b2e8f70be3f414d33e529975/rcl_lifecycle/test/test_rcl_lifecycle.cpp#L227-L238), `rcl` already depends on `rcl_interfaces`).

`rcl_lifecycle_state_machine_options_t` sounds better to me so you can include no-com related options there, though having them grouped separetly isn't a bad idea either ... I don't mind strongly.

Reply:
+1 for breaking API and switching to an options struct.

Reply:
So this turns out to become a bigger change than anticipated. 

While looking at the allocator, I think it's generally good practice to store the allocator used for `..._init` locally and reuse that allocator instance within the `..._fini`. However, the lifecycle state machine seems to receive an external allocator in its `fini`.
I'll address that accordingly as I won't be able to cleanly separate this PR from a potential follow-up PR which addresses the allocator.

Reply:
+1 for not taking an allocator in the fini.